### PR TITLE
Make filter config provider shared.

### PR DESCRIPTION
**Pull request by bianpengyuan**
This makes filter config provider shared between listeners, and change the mapping between filter config provider and subscription to 1 on 1.

fixes https://github.com/envoyproxy/envoy/issues/14934

Signed-off-by: Pengyuan Bian <bianpengyuan@google.com>

### CODE REVIEW

**Comment by bianpengyuan** [comment]
If this is on the right track, I will work on adding a test for this.

**Review by htuch — COMMENTED** [review]
@kyessenov can you take a pass on this one?
@bianpengyuan Thanks for the fix! This needs a regression test to be added. Also, something not clear to me is whether we have a more fundamental issue with ECDS warming that this is just patching over.
/wait

**Comment by htuch** [comment]
As per discussion in #14934 this might not be exactly the right fix.

**Comment by bianpengyuan** [comment]
@htuch ptal, thanks!

**Review by htuch — COMMENTED** [review]
This now seems the right solution, but I don't have cycles to do the low level review; my ECDS state only tracked up until fairly late PRs from @kyessenov but not after a bunch of iterations that were done between him and @mattklein123. @kyessenov @mattklein123 would one of you be able to take this? Thanks.

**Comment by mattklein123** [comment]
I can take a look.

**Comment by mattklein123** [comment]
Check CI please?

/wait

**Comment by bianpengyuan** [comment]
@mattklein123 done. Looks like the failures now are unrelated flakes?

File: test/per_file_coverage.sh [hunk 1]
```diff
@@ -9,8 +9,8 @@ declare -a KNOWN_LOW_COVERAGE=(
 "source/common/common/posix:94.1"
 "source/common/crypto:0.0"
 "source/common/event:93.4" # Emulated edge events guards don't report LCOV
-"source/common/filter:96.5"
-"source/common/filter/http:96.5"
+"source/common/filter:96.3"
+"source/common/filter/http:96.3"
```
Comment (mattklein123) [inline test/per_file_coverage.sh:13]: Is this change related to this PR?

File: source/common/filter/http/filter_config_discovery_impl.cc [hunk 2]
```diff
@@ -169,52 +174,43 @@ void FilterConfigSubscription::onConfigUpdateFailed(Config::ConfigUpdateFailureR
...
+    if (apply_without_warming) {
+      factory_context.initManager().add(provider->init_target_);
+    }
+    dynamic_filter_config_providers_.insert(
+        {provider_identifier, std::weak_ptr<DynamicFilterConfigProviderImpl>(provider)});
+    return provider;
+  } else {
+    auto existing_provider = it->second.lock();
+    ASSERT(existing_provider != nullptr,
+           absl::StrCat("Cannot find dynamic filter config provider ", filter_config_name));
```
Comment (mattklein123) [inline source/common/filter/http/filter_config_discovery_impl.cc:208]: This is just going to crash below so I would remove this ASSERT

File: source/common/filter/http/filter_config_discovery_impl.cc [hunk 3]
```diff
@@ -16,22 +16,29 @@ namespace Http {
...
+  subscription_->filter_config_provider_.emplace(this);
   tls_.set([](Event::Dispatcher&) { return std::make_shared<ThreadLocalConfig>(); });
 }
 
 DynamicFilterConfigProviderImpl::~DynamicFilterConfigProviderImpl() {
-  subscription_->filter_config_providers_.erase(this);
+  ASSERT(subscription_->filter_config_provider_.has_value());
+  subscription_->filter_config_provider_.reset();
+  // Remove the provider from the provider manager.
+  filter_config_provider_manager_.dynamic_filter_config_providers_.erase(provider_id_);
```
Comment (mattklein123) [inline source/common/filter/http/filter_config_discovery_impl.cc:41]: Given my recent work on callbacks I'm a little paranoid about logic like this. Is it guaranteed that the provider will be destroyed before the manager? Does the manger itself need to be locked to make sure it is still alive? I don't have my head fully wrapped around the lifetime of all of the different objects here. If it is guaranteed safe can you add more lifetime comments here and elsewhere?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks makes sense to me at a high level but a few questions and a request for more comments. Thank you!

/wait

File: source/common/filter/http/filter_config_discovery_impl.cc (hunk 3 above)
Comment (bianpengyuan) [reply source/common/filter/http/filter_config_discovery_impl.cc:41]: I followed RDS code to pass in a provider manager reference here, which I thought would be safe. I tried to track down the ownership of provider and provider manager and looks like both of them were owned by filter factory lambda: https://github.com/envoyproxy/envoy/blob/2509ee6159883d75f86a3af57fe0e6cc4b74bbdc/source/extensions/filters/network/http_connection_manager/config.cc#L167

Where `singletons` keeps a shared pointer of the provider manager, and `filter_config` keeps a unique pointer of the provider. I think the same lifecycle applies to RDS provider and provider manager. IIUC, this means that provider and manager destruction order is not deterministic? If so, to resolve this, we'd probably pass in the config provider manager shared pointer all the way through `HttpConnectionManagerConfig` to provider, instead of passing in a reference?

File: test/per_file_coverage.sh (hunk 1 above)
Comment (bianpengyuan) [reply test/per_file_coverage.sh]: The coverage used to drop a bit at a former commit. but the latest commit has 96.5% coverage rate. I reverted this back.

File: source/common/filter/http/filter_config_discovery_impl.cc (hunk 2 above)
Comment (bianpengyuan) [reply source/common/filter/http/filter_config_discovery_impl.cc]: Removed.

File: source/common/filter/http/filter_config_discovery_impl.cc (hunk 3 above)
Comment (bianpengyuan) [reply source/common/filter/http/filter_config_discovery_impl.cc:41]: > I think the good start is that adding a ASSERT(subscriptions_.empty()); in ~FilterConfigProviderManagerImpl

I added an assert.

File: include/envoy/filter/http/filter_config_provider.h [hunk 4]
```diff
@@ -15,7 +15,7 @@ namespace Http {
 using FilterConfigProvider =
     Envoy::Config::ExtensionConfigProvider<Server::Configuration::NamedHttpFilterConfigFactory,
                                            Envoy::Http::FilterFactoryCb>;
-using FilterConfigProviderPtr = std::unique_ptr<FilterConfigProvider>;
+using FilterConfigProviderPtr = std::shared_ptr<FilterConfigProvider>;
```
Comment (kyessenov) [inline include/envoy/filter/http/filter_config_provider.h:18]: nit: `FilterConfigProviderSharedPtr`

File: source/common/filter/http/filter_config_discovery_impl.cc [hunk 5]
```diff
@@ -169,52 +174,46 @@ void FilterConfigSubscription::onConfigUpdateFailed(Config::ConfigUpdateFailureR
...
-    factory_context.initManager().add(provider->init_target_);
+  const std::string provider_identifier =
+      absl::StrCat(MessageUtil::hash(config_source), ".", filter_config_name);
+  auto it = dynamic_filter_config_providers_.find(provider_identifier);
+  if (it == dynamic_filter_config_providers_.end()) {
+    auto subscription = std::make_shared<FilterConfigSubscription>(
+        config_source, filter_config_name, factory_context, stat_prefix);
+    // For warming, wait until the subscription receives the first response to indicate readiness.
+    // Otherwise, mark ready immediately and start the subscription on initialization. A default
+    // config is expected in the latter case.
```
Comment (kyessenov) [inline source/common/filter/http/filter_config_discovery_impl.cc:197]: Disclaimer: I haven't understood the entire change. It seems to me that subscriptions and filter config providers cannot be 1-1 since filter config provider may have different type_urls and warming settings. It's possible to have two providers that have different defaults, but subscribe to the same resource. How do we handle that here?

Comment (kyessenov) [reply source/common/filter/http/filter_config_discovery_impl.cc:197]: Thinking out loud: maybe we need to initialize unique config providers better from the same subscription? That way we can handle "do not warm if already sent" and "duplicate config pushed" cases better.

File: test/common/filter/http/filter_config_discovery_impl_test.cc [hunk 6]
```diff
@@ -263,6 +263,7 @@ TEST_F(FilterConfigDiscoveryImplTest, DualProviders) {
       TestUtility::decodeResources<envoy::config::core::v3::TypedExtensionConfig>(response);
   EXPECT_CALL(init_watcher_, ready());
   callbacks_->onConfigUpdate(decoded_resources.refvec_, response.version_info());
+  EXPECT_EQ(provider_.get(), provider2.get());
```
Comment (kyessenov) [inline test/common/filter/http/filter_config_discovery_impl_test.cc:266]: Do you think we can have a test that describes the bad case that this PR fixes?

File: source/common/filter/http/filter_config_discovery_impl.cc (hunk 5 above)
Comment (bianpengyuan) [reply source/common/filter/http/filter_config_discovery_impl.cc:197]: > It seems to me that subscriptions and filter config providers cannot be 1-1 since filter config provider may have different type_urls and warming settings. It's possible to have two providers that have different defaults, but subscribe to the same resource. How do we handle that here?

I thought about this a bit. `apply_without_warming` was handled, by adding subscription init target into init manager if `apply_without_warming=false`. Default config does not seem to be a concern for provider. It is dealt within `HttpConnectionManagerConfig`, one level above provider. I did not consider `require_type_urls`. If we are going with change as this PR, we probably could add it into provider id hash?

> maybe we need to initialize unique config providers better from the same subscription? That way we can handle "do not warm if already sent" and "duplicate config pushed" cases better.

Are you thinking about something like https://github.com/envoyproxy/envoy/pull/14985/commits/6c677a48702a966f6333ec689198f631892f8310 ?

**Comment by bianpengyuan** [comment]
Close in favor of https://github.com/envoyproxy/envoy/pull/15371
